lib/schnorr.py

```diff
@@ -221,3 +246,201 @@ def verify(pubkey, signature, message_hash):
+        # we store k in a list since .pop() is atomic.
+        self._kcontainer = [k]
+        Rpoint = k * ecdsa.SECP256k1.generator
+        self.R = point_to_ser(Rpoint, comp=True)
+
+    def get_R(self):
+        return self.R
+
+    def sign(self, privkey, ebytes):
+        assert len(privkey) == 32
+        assert len(ebytes) == 32
+        try:
+            k = self._kcontainer.pop()
+        except IndexError:
+            raise RuntimeError("Attempted to sign twice!")
+
+        x = int.from_bytes(privkey, 'big')
+        e = int.from_bytes(ebytes, 'big')
+
+        s = (k + e * x) % self.order
+        return s.to_bytes(32, 'big')
+
+
+class BlindSignatureRequest:
+    """ Schnorr blind signature creator, requester side.
+
+    We expect to be set up with two elliptic curve points
+    (serialized as bytes) -- the Blind signer's public key, and
+    a nonce point whose secret is known by the signer. Also, the
+    32-byte message_hash should be provided.
+
+    Upon construction, this creates and remembers the blinding factors,
+    and also performs the expensive math needed to create the blind
+    signature request. One initialized, call .get_request() to obtain
+    the 32-byte request that should be sent to the signer. Once you get
+    back their 32-byte response, call finalize().
+
+    The resultant Schnorr signatures follow the standard BCH Schnorr
+    convention (using Jacobi symbol, pubkey prefixing and SHA256).
+
+    Internally we use two random blinding factors a,b. Due to the jacobi
+    thing, we have to also include a signflip factor c = +/- 1.
+
+        [signer provides: R = k*G]
+        R' = c*(R + a*G + b*P)
+        choose c = +1 or -1 such that jacobi(R'.y(), fieldsize) = +1
+        e' = Hash(R'.x | ser_compressed(P) | message32)
+        e = c*e' + b mod n
+        [send to signer: e]
+        [signer provides: s = k + e*x]
+        s' = c*(s + a) mod n
+
+        resulting unblinded signature: (R'.x, s')
+
+    Ref: https://blog.cryptographyengineering.com/a-note-on-blind-signature-schemes/
+    """
+    order = ecdsa.SECP256k1.generator.order()
+    fieldsize = ecdsa.SECP256k1.curve.p()
+
+    def __init__(self, pubkey, R, message_hash):
+        """ Expects three bytes objects """
+        assert isinstance(pubkey, bytes)
+        assert isinstance(R, bytes)
+        assert len(message_hash) == 32
+
+        self.pubkey = pubkey
+        self.R = R
+        self.message_hash = message_hash
+
+        self.a = ecdsa.util.randrange(self.order)
+        self.b = ecdsa.util.randrange(self.order)
+        if seclib:
+            self._calc_initial_fast()
+        else:
+            self._calc_initial()
+        assert self.c in (-1, +1)
+        ehash = hashlib.sha256(self.Rxnew + self.pubkey_compressed + message_hash).digest()
+        self.e = (self.c * int.from_bytes(ehash,'big') + self.b) % self.order
+
+        self.enew = int.from_bytes(ehash,'big') % self.order # debug
```

Comment:
Wouldn't the crypto nerds get mad at you for using modulo here? (I seriously have no idea)

Reply:
Perhaps a bit yes, but that's how the Schnorr algorithm is defined. It turns out that since this `e` integer is not a secret value, and due to how Schnorr works, a very slight bit of bias is not hurting it.

Reply:
Ah ok. Fair enough. You did say "debug" there so I was like "maybe he forgot to finish something". 👍